### ISSUE TITLE
Issue with quantize_static

### ISSUE THREAD

**Issue by arpdubey** [microsoft/onnxruntime-inference-examples#319]
Accuracy for normal resnet50.onnx model is coming out to be above 70% but after quantizing it, accuracy becomes 0.10%.. What could be the issue?
Any help would be appreciated

**Comment by tianleiwu** [comment]
@arpdubey, could you share information about how you quantized the model, and provide evaluation script to reproduce the accuracy?

**Comment by arpdubey** [comment]
input_model_path = "resnet50.onnx"
output_model_path = "updated20.onnx"
calibration_dataset_path = "test_images/"
dr = resnet50_data_reader.ResNet50DataReader(calibration_dataset_path, input_model_path)       quantize_static(input_model_path,output_model_path,dr,per_channel=False,quant_format=QuantFormat.QOperator,weight_type=QuantType.QUInt8,activation_type=QuantType.QUInt8 )

**Comment by arpdubey** [comment]
Resnet50_data_reader
import numpy
import onnxruntime
import os
from onnxruntime.quantization import CalibrationDataReader
from PIL import Image


def _preprocess_images(images_folder: str, height: int, width: int, size_limit=0):
    """
    Loads a batch of images and preprocess them
    parameter images_folder: path to folder storing images
    parameter height: image height in pixels
    parameter width: image width in pixels
    parameter size_limit: number of images to load. Default is 0 which means all images are picked.
    return: list of matrices characterizing multiple images
    """
    image_names = os.listdir(images_folder)
    if size_limit > 0 and len(image_names) >= size_limit:
        batch_filenames = [image_names[i] for i in range(size_limit)]
    else:
        batch_filenames = image_names
    unconcatenated_batch_data = []

    for image_name in batch_filenames:
        image_filepath = images_folder + "/" + image_name
        pillow_img = Image.new("RGB", (width, height))
        pillow_img.paste(Image.open(image_filepath).resize((width, height)))
        input_data = numpy.float32(pillow_img) - numpy.array(
            [123.68, 116.78, 103.94], dtype=numpy.float32
        )
        nhwc_data = numpy.expand_dims(input_data, axis=0)
        nchw_data = nhwc_data.transpose(0, 3, 1, 2)  # ONNX Runtime standard
        unconcatenated_batch_data.append(nchw_data)
    batch_data = numpy.concatenate(
        numpy.expand_dims(unconcatenated_batch_data, axis=0), axis=0
    )
    return batch_data


class ResNet50DataReader(CalibrationDataReader):
    def __init__(self, calibration_image_folder: str, model_path: str):
        self.enum_data = None

        # Use inference session to get input shape.
        session = onnxruntime.InferenceSession(model_path, providers=['CPUExecutionProvider'])
        (_, _, height, width) = session.get_inputs()[0].shape

        # Convert image to input data
        self.nhwc_data_list = _preprocess_images(
            calibration_image_folder, height, width, size_limit=0
        )
        self.input_name = session.get_inputs()[0].name
        self.datasize = len(self.nhwc_data_list)

    def get_next(self):
        if self.enum_data is None:
            self.enum_data = iter(
                [{self.input_name: nhwc_data} for nhwc_data in self.nhwc_data_list]
            )
        return next(self.enum_data, None)

    def rewind(self):
        self.enum_data = None

**Comment by arpdubey** [comment]
import torch
from PIL import Image
import numpy as np
import onnxruntime
import os
import time
start_time=time.time()
from torchvision import transforms as transforms
onnx_model_path='uodated20.onnx'
session=onnxruntime.InferenceSession(onnx_model_path, providers=["CPUExecutionProvider"])
input_names = session.get_inputs()[0].name


def get_lable(filename): 
    filename = filename.replace('.JPEG', '')
    return " ".join(filename.split('_')[1:])
    
preprocess = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])
def accuracy():
    with open("imagenet_classes.txt", "r") as f:
        categories = [s.strip() for s in f.readlines()]
    image_dir = "/home/ext/arpdubey/imagenet-sample-images/"
    images = [filename for filename in os.listdir(image_dir) if filename.endswith('.JPEG')]
    totalImages = len(images)
    correctPrediction = 0
    # count=0
    for i in images:
        ground_truth = get_lable(i)
        
        image = Image.open(image_dir + i)
        try:
            if image.mode!='RGB':
                 image=image.convert('RGB')
            input_tensor = preprocess(image)
            input_tensor=input_tensor.unsqueeze(0)
            input_arr = input_tensor.cpu().detach().numpy()
            ort_outputs = session.run([], {input_names: input_arr})[0]
            output = ort_outputs.flatten()
            top5_catid = np.argsort(-output)[:5]
            if(ground_truth == categories[top5_catid[0]]):
                    correctPrediction+=1
                    print("hero")
        except RuntimeError:
             print("Error")
    print("Total Prediction: ",totalImages)
    print("Correct Prediction: ",correctPrediction)  
    # print("Failed Images Count: ", count)     
    return (correctPrediction/totalImages)

accuracy=accuracy()
print('Accuracy of the ONNX model on ImageNet dataset: {:.2%}'.format(accuracy))
end_time=time.time()
print("Time taken: ",end_time-start_time,"seconds")

**Comment by arpdubey** [comment]
`import torch
from PIL import Image
import numpy as np
import onnxruntime
import os
import time
start_time=time.time()
from torchvision import transforms as transforms
onnx_model_path='1.onnx'
session=onnxruntime.InferenceSession(onnx_model_path, providers=["CPUExecutionProvider"])
input_names = session.get_inputs()[0].name


def get_lable(filename): 
    filename = filename.replace('.JPEG', '')
    return " ".join(filename.split('_')[1:])
    
preprocess = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])
def accuracy():
    with open("imagenet_classes.txt", "r") as f:
        categories = [s.strip() for s in f.readlines()]
    image_dir = "/home/ext/arpdubey/imagenet-sample-images/"
    images = [filename for filename in os.listdir(image_dir) if filename.endswith('.JPEG')]
    totalImages = len(images)
    correctPrediction = 0
    # count=0
    for i in images:
        ground_truth = get_lable(i)
        
        image = Image.open(image_dir + i)
        try:
            if image.mode!='RGB':
                 image=image.convert('RGB')
            input_tensor = preprocess(image)
            input_tensor=input_tensor.unsqueeze(0)
            input_arr = input_tensor.cpu().detach().numpy()
            ort_outputs = session.run([], {input_names: input_arr})[0]
            output = ort_outputs.flatten()
            top5_catid = np.argsort(-output)[:5]
            if(ground_truth == categories[top5_catid[0]]):
                    correctPrediction+=1
                    print("hero")
        except RuntimeError:
             print("Error")
    print("Total Prediction: ",totalImages)
    print("Correct Prediction: ",correctPrediction)  
    # print("Failed Images Count: ", count)     
    return (correctPrediction/totalImages)

accuracy=accuracy()
print('Accuracy of the ONNX model on ImageNet dataset: {:.2%}'.format(accuracy))
end_time=time.time()
print("Time taken: ",end_time-start_time,"seconds")`

**Comment by arpdubey** [comment]
> @arpdubey, could you share information (like the quantization script) about how you quantized the model, and provide evaluation script to reproduce the accuracy?

Thanks for your response.. 
When I am quantizing nodes present in the other half of resnet model, it is giving me good accuracy while if I am quantizing starting nodes then accuracy drops down to 0%

**Comment by yufenglee** [comment]
how is your calibration data selected?

**Comment by arpdubey** [comment]
I have just randomly took it.. Any suggestions from you regarding selection of calibration data?

**Comment by ppppbamzy** [comment]
I had the same problem too, until I modified the normalization step of image preprocessing of [resnet50_data_reader.py](https://github.com/microsoft/onnxruntime-inference-examples/blob/afc83c003c2b262929d1667d8934021bfde8f394/quantization/image_classification/cpu/resnet50_data_reader.py):
```
diff --git a/quantization/image_classification/cpu/resnet50_data_reader.py b/quantization/image_classification/cpu/resnet50_data_reader.py
index d07198f..9ebd899 100644
--- a/quantization/image_classification/cpu/resnet50_data_reader.py
+++ b/quantization/image_classification/cpu/resnet50_data_reader.py
@@ -25,9 +25,8 @@ def _preprocess_images(images_folder: str, height: int, width: int, size_limit=0
         image_filepath = images_folder + "/" + image_name
         pillow_img = Image.new("RGB", (width, height))
         pillow_img.paste(Image.open(image_filepath).resize((width, height)))
-        input_data = numpy.float32(pillow_img) - numpy.array(
-            [123.68, 116.78, 103.94], dtype=numpy.float32
-        )
+        input_data = ((numpy.float32(pillow_img) / 255.0 - [0.485, 0.456, 0.406]) /
+                      [0.229, 0.224, 0.225]).astype(numpy.float32)
         nhwc_data = numpy.expand_dims(input_data, axis=0)
         nchw_data = nhwc_data.transpose(0, 3, 1, 2)  # ONNX Runtime standard
         unconcatenated_batch_data.append(nchw_data)
```
Does this mean the current code is incomplete?
------
BTW, there's [discussion](https://github.com/microsoft/onnxruntime-inference-examples/issues/262) about those magic numbers `[123.68, 116.78, 103.94]`, and it can be seen that they are just `[0.485, 0.456, 0.406] * 255.0`.
